### PR TITLE
S151 Github: Rules for proxy

### PR DESCRIPTION
**Pull request by aperez-worklytics**
- Added documentation about how to build the connector through Github Apps
- Added bunch of rules and tests for GH REST API endpoints (repos, logs, events, timeline, pull request, commits, comments, etc)
- Added a rule for querying the GraphQL only for users; as the endpoint is open to retrieve everything is done in a request, a response has been created to include only the stuff that we are going to query.

NOTE: Most of the lines included in the PR are related to the files and sanitized files that it contains.

### Features
[Psoxy rules support](https://app.asana.com/0/0/1204953309467291)

### Change implications

 - dependencies added/changed? **no**

### CODE REVIEW

File: docs/sources/api-response-examples/github/sanitized/graph_api_users.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            "hasNextPage":false,
+            "endCursor":"Y3Vyc29yOnYyOpHOAMl1tg=="
+          },
+          "edges":[
+            {
+              "node":{
+                "samlIdentity":{
+                  "nameId":"{\"scope\":\"email\",\"domain\":\"contoso.com\",\"hash\":\"ui7ZfyKC2nursXtsx_lyORcptfTrCuGf38xextkHyW8\"}"
+                },
+                "guid":"abc3f758-1a88-11ee-85d2-07ac5bf201d6",
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/graph_api_users.json:16]: does this need to be pseudonymized?

File: docs/sources/api-response-examples/github/sanitized/graph_api_users.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                  "login":"{\"scope\":\"github\",\"hash\":\"iK5cxa4JLZ9qQnICuYKRS8wJoRkdvXv6xrLAIg2.ETs\"}"
+                }
+              }
+            },
+            {
+              "node":{
+                "samlIdentity":{
+                  "nameId":"{\"scope\":\"email\",\"domain\":\"contoso.com\",\"hash\":\"qyfnJEcnYW8yjJaykmCWONL176MJJardThqpzjVwkiM\"}"
+                },
+                "guid":"acbe9c8e-1a99-11ee-85b8-61109bf19007",
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/graph_api_users.json:27]: also?

File: docs/sources/api-response-examples/github/sanitized/org_audit_log.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    "operation_type":"modify",
+    "org":"some-org",
+    "org_id":137521429,
+    "public_repo":false,
+    "pull_request_id":1423311422,
+    "pull_request_url":"https://github.com/some-org/test/pull/1",
+    "repo":"some-org/test",
+    "repository":"some-org/test",
+    "user":"{\"scope\":\"github\",\"hash\":\"wzPpLnb2eF.USVQDBlxB2HBvVi1dsgdwIJdnEIUU8Pw\"}",
+    "user_id":13968776
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/org_audit_log.json:51]: sanitize?

File: docs/sources/api-response-examples/github/sanitized/org_audit_log.json [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    "actor_location":{
+      "country_code":"GB"
+    },
+    "org":"some-org",
+    "org_id":137521429,
+    "repo":"some-org/test",
+    "repository":"some-org/test",
+    "repository_public":false,
+    "user":"{\"scope\":\"github\",\"hash\":\"7mRFMiuvI7YeApDZjGtBhVAh3kuQdFoz7HqJKGIr674\"}",
+    "user_id":0
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/org_audit_log.json:30]: sanitize?

File: docs/sources/api-response-examples/github/sanitized/org_audit_log.json [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/org_audit_log.json:15]: sanitize?

File: docs/sources/api-response-examples/github/sanitized/repo_comments.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
+[
+  {
+    "html_url": "https://github.com/octocat/Hello-World/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e#commitcomment-1",
+    "url": "https://api.github.com/repos/octocat/Hello-World/comments/1",
+    "id": 1,
+    "node_id": "MDEzOkNvbW1pdENvbW1lbnQx",
+    "body": "Great stuff",
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/repo_comments.json:7]: should be dropped

File: docs/sources/api-response-examples/github/sanitized/repo_comments.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    "path": "file1.txt",
+    "position": 4,
+    "line": 14,
+    "commit_id": "6dcb09b5b57875f334f61aebed695e2e4193db5e",
+    "user": {
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/repo_comments.json:12]: user appears not be sanitized

File: docs/sources/api-response-examples/github/sanitized/repos.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    "id":408943403,
+    "node_id":"R_kgDOGF_7Kw",
+    "name":"psoxy",
+    "full_name":"Worklytics/psoxy",
+    "private":false,
+    "owner":{
+      "login":"{\"scope\":\"github\",\"hash\":\"2sH2Fo.Zu09PBKAf5I4pHaoNy5uvcZ3zWs9SiWUBTWQ\"}",
+      "id":"{\"scope\":\"github\",\"hash\":\"VEvLkj8lIGM8HV6s3dhWbUNVGH0v8cK3LYhYxx5DcsE\"}",
+      "node_id":"{\"scope\":\"github\",\"hash\":\"NHcnDQoWyKLZxrrtTxT9AJKyPbhc_0Ju7ynKtCTwI_c\"}",
+      "type":"Organization",
```
Comment (eschultink) [inline docs/sources/api-response-examples/github/sanitized/repos.json:12]: i think we can in principal skip sanitizing owner where type == Organization, if that's possible.  Or just `owner` on the repos endpoint probably, as in our use case don't expect it to be a user.

File: java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+                    .jsonPath("$..labels_url")
+                    .jsonPath("$..comments_url")
+                    .jsonPath("$..events_url")
+                    .jsonPath("$..html_url")
+                    .jsonPath("$..title")
+                    .jsonPath("$..body")
+                    .jsonPath("$..description")
+                    .jsonPath("$..name")
+                    .jsonPath("$..diff_url")
+                    .jsonPath("$..patch_url")
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java:175]: why redact all the URLs? don't seem inherently unsafe to me. concern that we might see a repo owned by an individual or something, then get a github user name?

File: java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+                    .jsonPath("$..subscription_url")
+                    .jsonPath("$..tags_url")
+                    .jsonPath("$..teams_url")
+                    .jsonPath("$..trees_url")
+                    .jsonPath("$..clone_url")
+                    .jsonPath("$..mirror_url")
+                    .jsonPath("$..hooks_url")
+                    .jsonPath("$..svn_url")
+                    .jsonPath("$..homepage")
+                    .jsonPath("$..homepage")
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java:317]: duplicated?

File: java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
```
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java:386]: again, duplicated?

File: java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java
Comment (eschultink) [inline java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java:1]: generally - I think not necessary or useful to redact all the URLs, unless we expect them to reference individuals. 

I think that 
1) we can expect all the "repos" we get to be org ones,
2) urls desirable for many applications where we might want to be able to link back to the original content, 
3) cases where url is ambiguous - may refer to organization or user - I think it's usual and fair to presume it's an organization
4) if we really think there's a case where we expect a lot of references by url to users, then we should use a transformation to reversibly pseudonymize just the username portion of the url, imho.

**Review by eschultink — CHANGES_REQUESTED** [review]
👍  a few suggestions on the rules; the only blockers are a handful of cases where seems like user_id comes through, and one `user` object  - then general suggestion that i don't think need to be as aggressive redacting `_url_` fields everywhere

File: java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java
Comment (aperez-worklytics) [reply java/core/src/main/java/co/worklytics/psoxy/rules/github/PrebuiltSanitizerRules.java:1]: >> if we really think there's a case where we expect a lot of references by url to users, then we should use a transformation to reversibly pseudonymize just the username portion of the url, imho.

Ok; all rules coming from users are pseudonymized EXCEPT the ones from owner in repo, following your suggestion. It is true that during my tests, my repo has a user as an owner and not an org -because it was created by me, not by my org-.

**Review by aperez-worklytics — COMMENTED** [review]
Thanks! feedback applied

File: docs/sources/api-response-examples/github/sanitized/repos.json [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  {
+    "id":408943403,
+    "node_id":"R_kgDOGF_7Kw",
+    "name":"psoxy",
+    "full_name":"Worklytics/psoxy",
+    "private":false,
+    "owner":{
+      "login":"Worklytics-user",
+      "id":23456789,
+      "node_id":"MDEyOk9yZ2FuaXphdGlvbjE5OTc4MDk0",
```
Comment (jlorper) [inline docs/sources/api-response-examples/github/sanitized/repos.json:11]: are these 3 related to users somehow?

File: infra/examples-dev/aws-all/main.tf [hunk 13]
```diff
@@ -34,11 +34,13 @@ module "worklytics_connectors" {
   source = "../../modules/worklytics-connectors"
   # source = "git::https://github.com/worklytics/psoxy//infra/modules/worklytics-connectors?ref=rc-v0.4.29"
 
-  enabled_connectors    = var.enabled_connectors
-  example_jira_issue_id = var.example_jira_issue_id
-  jira_cloud_id         = var.jira_cloud_id
-  jira_server_url       = var.jira_server_url
-  salesforce_domain     = var.salesforce_domain
+  enabled_connectors        = var.enabled_connectors
+  example_jira_issue_id     = var.example_jira_issue_id
```
Comment (jlorper) [inline infra/examples-dev/aws-all/main.tf:38]: Disregard for this PR, just a general comment for future: but I see this growing and find it painful. To me, we should have a map (string, object)
```tf
enabled_connectors_settings = {
"jira": {
"jira_issue_id" : var.example_jira_issue_id
}
```
or
```tf
enabled_connectors_settings = {
jira_example_issue_id = var.example_jira_issue_id,
jira_cloud_id = ...
}
```
and maybe rely on local validators on what to expect

File: infra/examples-dev/aws-google-workspace/main.tf [hunk 14]
```diff
@@ -85,6 +85,8 @@ module "psoxy" {
   salesforce_domain              = var.salesforce_domain
   jira_server_url                = var.jira_server_url
   jira_cloud_id                  = var.jira_cloud_id
+  github_organization            = var.github_organization
+  example_github_repository      = var.example_github_repository
```
Comment (jlorper) [inline infra/examples-dev/aws-google-workspace/main.tf:89]: we should prefix everything by connector, easier to group visually and find:
```
```
```suggestion
  github_organization            = var.github_organization
  github_example_repository      = var.example_github_repository
```
and move up `example_jira_issue_id` with the other jira's.
If they were as I said you wouldn't have splitted them :smile:

**Review by jlorper — APPROVED** [review]
OK, some suggestions, no blockers

**Review by eschultink — APPROVED** [review]

File: infra/examples-dev/aws-all/main.tf (hunk 13 above)
Comment (eschultink) [reply infra/examples-dev/aws-all/main.tf:38]: https://app.asana.com/0/1201039336231823/1205054651201385

File: docs/sources/api-response-examples/github/sanitized/graph_api_error.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
+{
+  "data":{
+    "organization":{
+      "samlIdentityProvider":null
+    }
+  },
+  "errors":[
+    {
+      "path":[
```
Comment (jlorper) [inline docs/sources/api-response-examples/github/sanitized/graph_api_error.json:9]: why the order of the attributes is changed? Shouldn't it maintain same order? Or lead to flaky tests?

**Comment by eschultink** [comment]
i'm supportive of merging this to rc-v0.4.30 whenever you're ready.

File: docs/sources/api-response-examples/github/sanitized/graph_api_error.json (hunk 15 above)
Comment (aperez-worklytics) [reply docs/sources/api-response-examples/github/sanitized/graph_api_error.json:9]: Fixed on https://github.com/Worklytics/psoxy/pull/429
